lean4-infoview/src/infoview/goals.tsx

```diff
@@ -78,33 +59,98 @@ function getFilteredHypotheses(hyps: InteractiveHypothesisBundle[], filter: Goal
     }, [])
 }
 
+interface HypProps {
+    hyp: InteractiveHypothesisBundle
+    mvarId?: MVarId
+}
+
+function Hyp({ hyp: h, mvarId }: HypProps) {
+    const locs = React.useContext(LocationsContext)
+
+    const namecls: string = 'mr1 ' +
+        (h.isInserted ? 'inserted-text ' : '') +
+        (h.isRemoved ? 'removed-text ' : '')
+
+    const names = InteractiveHypothesisBundle_nonAnonymousNames(h).map((n, i) =>
+        <span className={namecls + (isInaccessibleName(n) ? 'goal-inaccessible ' : '')} key={i}>
+            <SelectableLocation
+                locs={locs}
+                loc={mvarId && h.fvarIds && h.fvarIds.length > i ?
+                    { mvarId, loc: { hyp: h.fvarIds[i] }} :
+                    undefined
+                }
+                alwaysHighlight={false}
+            >{n}</SelectableLocation>
+        </span>)
+
+    const typeLocs: Locations | undefined = React.useMemo(() =>
+        locs && mvarId && h.fvarIds && h.fvarIds.length > 0 ?
+            { ...locs, subexprTemplate: { mvarId, loc: { hypType: [h.fvarIds[0], ''] }}} :
+            undefined,
+        [locs, mvarId, h.fvarIds])
+
+    const valLocs: Locations | undefined = React.useMemo(() =>
+        h.val && locs && mvarId && h.fvarIds && h.fvarIds.length > 0 ?
+            { ...locs, subexprTemplate: { mvarId, loc: { hypValue: [h.fvarIds[0], ''] }}} :
+            undefined,
+        [h.val, locs, mvarId, h.fvarIds])
+
+    return <div>
+        <strong className="goal-hyp">{names}</strong>
+        :&nbsp;
+        <LocationsContext.Provider value={typeLocs}>
```

Comment:
I had to look this up: apparently context changes cause rerenders even inside `useMemo`, so this is all kosher.

Reply:
Do you mean `React.memo`? But yeah, `useContext` inside `memo` is updated as appropriate.

Reply:
Yes, that's what I meant.